### PR TITLE
removed an incorretc return that caused missing API timing metrics for 4xx and 5xx errors

### PR DESCRIPTION
**Pull request by harshilsharma63**
#### Summary
Removed an incorrect return statement that caused missed timing metrics for API calls that resulted in 4xx and 5xx errors.

#### Ticket Link
Fixes https://mattermost.atlassian.net/browse/MM-60269

#### Release Note
```release-note
None
```

### CODE REVIEW

**Comment by harshilsharma63** [comment]
Hi @amyblais , this needs to be cherrypicked into v10 as it block the release.

**Review by agnivade — CHANGES_REQUESTED** [review]
This still won't log the metrics if it hits the first condition of exceeding the URL length right?

I think the correct solution here would be to keep using the return statements, but move the metrics computation code inside the defer block declared in L169. Then the rest of the code remains easy to read without any inconsistencies.

Additionally, let's also add a unit test to lock in the functionality.

**Comment by harshilsharma63** [comment]
> This still won't log the metrics if it hits the first condition of exceeding the URL length right?
> 
> I think the correct solution here would be to keep using the return statements, but move the metrics computation code inside the defer block declared in L169. Then the rest of the code remains easy to read without any inconsistencies.
> 
> Additionally, let's also add a unit test to lock in the functionality.

Do we want to measure the time of an API call that we're rejecting on the very first line of logic? That URL length rejection have been in place for months BTW, its not new. I think we should not measure those, but your call.

**Comment by harshilsharma63** [comment]
@agnivade I've updated the code to record metrics for long URLs as well. We're not recording for APIs that exceed API rate limits and that will be handled separately as discussed.

File: server/channels/web/handlers.go
Comment (agnivade) [inline server/channels/web/handlers.go:187]: Actually, just a minor bug here. The `statusCode` variable needs to be set for `if statusCode != 0` check to work. And that used to work before because this got set in L386. 

Something like this could work:

```diff
--- a/server/channels/web/handlers.go
+++ b/server/channels/web/handlers.go
@@ -176,6 +176,7 @@ func (h Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
                if c.AppContext.Session() != nil {
                        responseLogFields = append(responseLogFields, mlog.String("user_id", c.AppContext.Session().UserId))
                }
+               statusCode = strconv.Itoa(w.(*responseWriterWrapper).StatusCode())
                // Websockets are returning status code 0 to requests after closing the socket
                if statusCode != "0" {
                        responseLogFields = append(responseLogFields, mlog.String("status_code", statusCode))
@@ -183,7 +184,7 @@ func (h Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
                mlog.Debug("Received HTTP request", responseLogFields...)

                if !rateLimitExceeded {
-                       h.recordMetrics(c, w, r, now)
+                       h.recordMetrics(c, r, now, statusCode)
                }
        }()

@@ -392,8 +393,7 @@ func (h Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
        }
 }

-func (h Handler) recordMetrics(c *Context, w http.ResponseWriter, r *http.Request, now time.Time) {
-       statusCode := strconv.Itoa(w.(*responseWriterWrapper).StatusCode())
+func (h Handler) recordMetrics(c *Context, r *http.Request, now time.Time, statusCode string) {
        if c.App.Metrics() != nil {
                c.App.Metrics().IncrementHTTPRequest()
```

**Review by agnivade — CHANGES_REQUESTED** [review]

File: server/channels/web/handlers.go
Comment (harshilsharma63) [reply server/channels/web/handlers.go]: Gotcha. Since the `statuscode` variable was being used only inside the defer block, I moved it completely in there.

**Review by agnivade — APPROVED** [review]
Nicely done. Thanks.

**Review by agarciamontoro — APPROVED** [review]
Oooh, I like the solution, thanks!

**Comment by agarciamontoro** [comment]
Tested that this works: the `mattermost_api_time` metric contains a comparable number of errors as the `loadtest_http_errors_total`.

image

There is a difference in the case of status_code == 403 , though. In that case, it's the `loadtest_http_errors_total` metric the one that doesn't register the datapoint. However, I was able to reproduce this with Mattermost v9.11, so it's not a new regression, and in any case a bug in the load-test tool, not in Mattermost:

image

I'll merge now.

**Comment by agnivade** [comment]
> However, I was able to reproduce this with Mattermost v9.11, so it's not a new regression, and in any case a bug in the load-test tool, not in Mattermost:

Do we need a ticket for this @agarciamontoro ?

**Pull request merged**
